File: coin-change/coin-change.java
class Solution {
    public static final int INV = 2000000000;
    public int coinChange(int[] coins, int amount) {
        if (coins == null || coins.length == 0)
            return -1;
        if (amount == 0)
            return 0;
        
        int[][] mat = new int[coins.length][amount+1];
        // Set up first row (base case)
        for (int i = 1; i < amount + 1; i++) {
            mat[0][i] = coins[0] > i ? mat[0][i] = INV : mat[0][i - coins[0]] + 1;
        }
        
        // Create the rest of the matrix
        for (int row = 1; row < coins.length; row++) {
            for (int change = 1; change < amount+1; change++) {
                mat[row][change] = coins[row] > change ? Math.min(INV, mat[row-1][change]) : Math.min(mat[row-1][change], 1 + mat[row][change-coins[row]]);
                }
            }

        return mat[coins.length-1][amount] >= 1000000000 ? -1 : mat[coins.length-1][amount];
    }
}

File: coin-change/coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        INV = 2_000_000_000
        if coins == None or len(coins) == 0:
            return -1
        if amount == 0:
            return 0
        dp = [[0]*(amount+1) for i in range(len(coins))]
        for i in range(1, amount+1):
            dp[0][i] = INV if coins[0] > i else dp[0][i-coins[0]] + 1
            
        for row in range(1, len(coins)):
            for change in range(1, amount+1):
                dp[row][change] = min(INV, dp[row-1][change]) if coins[row] > change else min(dp[row-1][change], 1 + dp[row][change-coins[row]])
        return -1 if dp[len(coins) - 1][amount] >= 1_000_000_00 else dp[len(coins) - 1][amount]
                
        